### PR TITLE
Enable mypy check for distributed.py, and fix type errors

### PR DESCRIPTION
**Pull request by yhcharles**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #87543

### CODE REVIEW

File: torch/_C/_distributed_c10d.pyi [hunk 1]
```diff
@@ -27,18 +28,48 @@ class GradBucket:
...
         params: List[Tensor],
         bucket_indices: List[List[int]],
+        per_bucket_size_limits: List[int],
         process_group: ProcessGroup,
-        expect_sparse_gradients: List[bool],
-        bucket_bytes_cap: int,
-        find_unused_parameters: bool,
-        gradient_as_bucket_view: bool,
+        expect_sparse_gradients: List[bool] = [],
+        bucket_bytes_cap: int = kDefaultBucketBytesCap,
```
Comment (ezyang) [inline torch/_C/_distributed_c10d.pyi:46]: instead of having to hardcode the constant here (where it can diverge), you can use an ellipsis instead

**Review by ezyang — COMMENTED** [review]
Thanks! You might want to get someone from distributed to look over too

File: torch/_C/_distributed_c10d.pyi [hunk 2]
```diff
@@ -1,8 +1,9 @@
 from datetime import timedelta
 from enum import Enum
-from typing import Optional, List, Any, Tuple, overload, Union
+from typing import Dict, Optional, List, Any, Tuple, overload, Union
```
Comment (fduwjj) [inline torch/_C/_distributed_c10d.pyi:3]: Nit: ideally we should sort it alphabetically?

File: torch/nn/parallel/distributed.py [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import sys
 import copy
 from dataclasses import dataclass
-from typing import Callable, Any, Type
+from typing import Callable, Any, Optional, Type
```
Comment (fduwjj) [inline torch/nn/parallel/distributed.py:4]: ditto.

**Review by fduwjj — APPROVED** [review]
lgtm.

File: torch/_C/_distributed_c10d.pyi (hunk 2 above)
Comment (yhcharles) [reply torch/_C/_distributed_c10d.pyi]: fixed

File: torch/nn/parallel/distributed.py (hunk 3 above)
Comment (yhcharles) [reply torch/nn/parallel/distributed.py]: fixed

**Pull request closed without merging**